Don't use docker.Client instance from context if missing attributes

terminalmage opened the pull request:
Under some circumstances (which I can't reliably reproduce), the
docker.Client() instance does not have a timeout attribute, causing
anything that needs to use the docker client to result in a traceback.

This fixes this corner case (later branches initialize the timeout
differently).

salt-jenkins commented:
@terminalmage, thanks for your PR! By analyzing the history of the files in this pull request, we identified @ticosax, @rallytime and @tyhunt99 to be potential reviewers.

terminalmage commented:
Hold off on merging this, the issue might be something with docker-py itself.

terminalmage committed:
Don't use docker.Client instance from context if missing attributes

Under some circumstances (which I can't reliably reproduce), the
docker.Client() instance does not have a timeout attribute, causing
anything that needs to use the docker client to result in a traceback.

This fixes this corner case (later branches initialize the timeout
differently).

terminalmage commented:
OK I changed the approach on this PR and renamed it. For some reason, on one of @SaltyCharles' test boxes, the client instance when returned from the ``__context__`` dunder has no timeout, base_url, and probably other attributes. Thing is, you can instantiate a client instance in a docker shell and it has a timeout attribute, so why the instance returned from the ``__context__`` dunder is missing one I do not know. There may be something else interfering that I'm not seeing. Regardless, I've worked around this odd corner case by generating a new Client instance if the one in the dunder lacks the ``timeout`` attribute. It's a weird, hacky fix, but I honestly can't think of a more elegant solution since there doesn't appear to be any self-evident reason behind the weird behavior.

ticosax commented:
@terminalmage Can you give some instructions so I can reproduce this behavior myself ?

terminalmage commented:
@ticosax Sorry, but I can't. I only ever reproduced this on one of @SaltyCharles' VMs, which he gave me access to use. I tried to reproduce on my laptop, and on a couple Vagrant boxes with Docker installed on them, and I never was able to reproduce.

The PR should be pretty straightforward though.

ticosax commented:
It might be a bigger issue than docker-py itself, or just an edge case related to local setup of packages.
In any case I'd put a comment around that line explaining what is going on.
I'll be able to submit a PR in couple of days.

terminalmage commented:
@ticosax See #39988, I added a comment.
